STM/vertx/README.adoc

```diff
@@ -0,0 +1,179 @@
+
+What it's good for:
+
+  - horizontal scaling by using better hardware so that more threads can do be used to service the
+    workload;
+
+Usage:
+------
+
+  mvn exec:java -Pdemo2
+
+or
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo12.PersistentSharedTheatreVerticle
+
+Start a second verticle that will share the same STM object. In the output of the above command
+look for the line that reports the uid of the STM object, for example:
+
+> Theatre STM uid: 0:ffffac1182c6:9197:5912d4ff:1
+
+The actual uid may change from run to run. Pass this uid the second JVM using a java system property:
+
+  mvn exec:java -Pdemo2 -Duid=0:ffffac1182c6:9197:5912d4ff:1
+
+or
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo12.PersistentSharedTheatreVerticle 0:ffffac1182c6:9197:5912d4ff:1
+
+Create two bookings using services running in the two different JVMs (on endpoints 8080 and 8082
+respectively):
+
+  curl -X POST http://localhost:8080/api/theatre/Apollo
+  curl -X POST http://localhost:8082/api/theatre/Savoy
+
+Check that each JVM reports the correct number of bookings (namely 2):
+
+  curl -X GET http://localhost:8080/api/theatre
+  curl -X GET http://localhost:8082/api/theatre
+
+Use Case 3: Managing shared state across different STM objects
+===============================================
+
+Characteristics:
+  - RECOVERABLE and EXCLUSIVE
+  - trip service verticle (multiple instances) using STM objects to maintain theatre and taxi bookings.
+  - providing HTTP endpoints for making trip, theatre and taxi bookings.
+
+What it's good for:
+  - composing transactional operations across different STM objects 
+
+The trip service fulfils booking requests by updating shared STM objects representing the theatre and
+taxi booking services respectively.
+
+  mvn exec:java -Pdemo3
+
+or
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo3.TripSTMVerticle
+
+Make two trip bookings:
+
+  curl -X POST http://localhost:8080/api/trip/Savoy/ABC
+  curl -X POST http://localhost:8080/api/trip/Apollo/XYZ
+
+and a single theatre booking:
+
+  curl -X POST http://localhost:8080/api/theatre/Savoy
+
+observe that each booking is serviced by a different verticle. Check that number of theatre (3) and
+number of taxi (2) bookings are correct:
+
+  curl -X GET http://localhost:8080/api/theatre
+  curl -X GET http://localhost:8080/api/taxi
+
+Use Case 4: Stress Testing:
+--------------------------
+
+Non of the previous use cases demonstrate contention of the STM objects. This next case will start
+use case 1 (theatre booking example) and then make lots of concurrent trip bookings and validate that
+the expected number of bookngs are made.
+
+Usage:
+------
+
+Start the theatre service in one window or in the background:
+
+  mvn exec:java -Pdemo1
+
+and now make lots of concurrent bookings:
+
+  mvn exec:java -Pstress
```

Comment:
It might be something in the phase that is being run (default-cli) because when I pass it on the command line it works:

$ mvn exec:java -Pstress -Dexec.cleanupDaemonThreads=false
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building Narayana - STM and vertx demo 5.6.0.Final-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- exec-maven-plugin:1.4.0:java (default-cli) @ stm-vertx-demo ---
Waiting for 1000 requests
0 out of 1000 requests failed in 7342 ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.743 s
[INFO] Finished at: 2017-05-11T12:00:13+01:00
[INFO] Final Memory: 15M/208M
[INFO] ------------------------------------------------------------------------